### ISSUE TITLE
Inconsistent behavior when accessing using go S3 sdk

### ISSUE THREAD

**Issue by arturo-myota** [localstack/localstack#3726]
# Type of request: This is a ...

[ x] bug report
[ ] feature request

# Detailed description
Using go sdk github.com/aws/aws-sdk-go v1.29.1

I can list objects stored in S3 using the go sdk github.com/aws/aws-sdk-go v1.29.1
Attempting to download them fails regardless of how the object's key is specified.
If the same key returned from ListObjects is used, then a download of that key returns 404.

If the key is URL encoded (either path encoded or query encoded) an error is printed and the return status is 500:
```
"error":"InternalServerError: Internal Server Error\n\tstatus code: 500, request id: , host id: "

2021-03-15T02:04:20:WARNING:bootstrap.py: Thread run method <function AdaptiveThreadPool.submit.<locals>._run at 0x7f8a47cf3d30>(None) failed: An error occurred (404) when calling the HeadObject operation: Not Found Traceback (most recent call last):
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 653, in run
    result = self.func(self.params)
  File "/opt/code/localstack/localstack/utils/async_utils.py", line 28, in _run
    return fn(*args, **kwargs)
  File "/opt/code/localstack/localstack/services/generic_proxy.py", line 607, in handler
    response = modify_and_forward(method=method, path=path_with_params, data_bytes=data, headers=headers,
  File "/opt/code/localstack/localstack/services/generic_proxy.py", line 373, in modify_and_forward
    listener_result = listener.forward_request(method=method,
  File "/opt/code/localstack/localstack/services/edge.py", line 105, in forward_request
    return do_forward_request(api, method, path, data, headers, port=port)
  File "/opt/code/localstack/localstack/services/edge.py", line 116, in do_forward_request
    result = do_forward_request_inmem(api, method, path, data, headers, port=port)
  File "/opt/code/localstack/localstack/services/edge.py", line 136, in do_forward_request_inmem
    response = modify_and_forward(method=method, path=path, data_bytes=data, headers=headers,
  File "/opt/code/localstack/localstack/services/generic_proxy.py", line 443, in modify_and_forward
    updated_response = update_listener.return_response(**kwargs)
  File "/opt/code/localstack/localstack/services/s3/s3_listener.py", line 1349, in return_response
    fix_range_content_type(bucket_name, path, headers, response)
  File "/opt/code/localstack/localstack/services/s3/s3_listener.py", line 524, in fix_range_content_type
    result = s3_client.head_object(Bucket=bucket_name, Key=key_name)
  File "/opt/code/localstack/.venv/lib/python3.8/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/opt/code/localstack/.venv/lib/python3.8/site-packages/botocore/client.py", line 676, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (404) when calling the HeadObject operation: Not Found
```

## Expected behavior
A key returned from List-Object should be successfully downloaded with Get-Object using the go sdk.

## Actual behavior

# Steps to reproduce

## Command used to start LocalStack
Starting localstack on darwin
Darwin local.local 19.6.0 Darwin Kernel Version 19.6.0: Tue Jan 12 22:13:05 PST 2021; root:xnu-6153.141.16~1/RELEASE_X86_64 x86_64
```
**$ docker run --rm -p 4566:4566 -p 4571:4571 -e SERVICES=s3  localstack/localstack**
Waiting for all LocalStack services to be ready
2021-03-15 01:47:40,509 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
2021-03-15 01:47:40,512 INFO supervisord started with pid 15
2021-03-15 01:47:41,484 INFO spawned: 'dashboard' with pid 21
2021-03-15 01:47:41,487 INFO spawned: 'infra' with pid 22
2021-03-15 01:47:41,492 INFO success: dashboard entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
2021-03-15 01:47:41,492 INFO exited: dashboard (exit status 0; expected)
(. .venv/bin/activate; exec bin/localstack start --host)
2021-03-15 01:47:42,496 INFO success: infra entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
Starting local dev environment. CTRL-C to quit.

LocalStack version: 0.12.7
LocalStack build date: 20210310
LocalStack build git hash: 95f91f6

Starting edge router (https port 4566)...
Starting mock S3 service on http port 4566 ...
2021-03-15T01:47:46:INFO:localstack.utils.analytics.profiler: Execution of "load_plugin_from_path" took 3793.308973312378ms
2021-03-15T01:47:46:INFO:localstack.utils.analytics.profiler: Execution of "load_plugins" took 3794.140100479126ms
2021-03-15T01:47:46:INFO:localstack.multiserver: Starting multi API server process on port 44181
[2021-03-15 01:47:46 +0000] [23] [INFO] Running on https://0.0.0.0:4566 (CTRL + C to quit)
2021-03-15T01:47:46:INFO:hypercorn.error: Running on https://0.0.0.0:4566 (CTRL + C to quit)
[2021-03-15 01:47:46 +0000] [23] [INFO] Running on http://0.0.0.0:44181 (CTRL + C to quit)
2021-03-15T01:47:46:INFO:hypercorn.error: Running on http://0.0.0.0:44181 (CTRL + C to quit)
Waiting for all LocalStack services to be ready
Waiting for all LocalStack services to be ready
Ready.
```

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)
The aws CLI (not awslocal) can list and return the objects successfully.
```
aws --endpoint-url=http://localhost:4566 s3api list-objects --bucket b
aws --endpoint-url=http://localhost:4566 s3api get-object --bucket b --key 6e74f92ddf74438a711d75e055
```
I can put together a test case in a few days.  Mostly reaching out to see if there's anything obvious I am missing.

**Comment by macnev2013** [comment]
Hi @arturo-myota , Thanks for reporting. It would be great if you can provide a small test case for the same. Looking forward to it. Thanks.

**Comment by whummer** [comment]
Thanks @arturo-myota @macnev2013 . This should be fixed in #3800 - can you please pull the `latest` Docker image and give it another try? Please report here if the problem persists. Thanks!

**Comment by arturo-myota** [comment]
Sorry for the late feedback.

I just tried this agains 0.12.15.  Still getting an error (I removed the actual URL for confidentiality):
`$ docker run --rm -p 4566:4566 -p 4571:4571 -e SERVICES=s3 localstack/localstack
Waiting for all LocalStack services to be ready
2021-07-14 17:46:09,639 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
2021-07-14 17:46:09,641 INFO supervisord started with pid 14
2021-07-14 17:46:10,649 INFO spawned: 'dashboard' with pid 20
2021-07-14 17:46:10,651 INFO spawned: 'infra' with pid 21
2021-07-14 17:46:10,656 INFO success: dashboard entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
2021-07-14 17:46:10,657 INFO exited: dashboard (exit status 0; expected)
(. .venv/bin/activate; exec bin/localstack start --host)
Starting local dev environment. CTRL-C to quit.
2021-07-14 17:46:12,409 INFO success: infra entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)

LocalStack version: 0.12.15
LocalStack build date: 2021-07-13
LocalStack build git hash: 0a3d5873

Starting edge router (https port 4566)...
Starting mock S3 service on http port 4566 ...
2021-07-14T17:46:15:INFO:localstack.multiserver: Starting multi API server process on port 54333
[2021-07-14 17:46:15 +0000] [22] [INFO] Running on https://0.0.0.0:4566 (CTRL + C to quit)
[2021-07-14 17:46:15 +0000] [22] [INFO] Running on http://0.0.0.0:54333 (CTRL + C to quit)
2021-07-14T17:46:15:INFO:hypercorn.error: Running on http://0.0.0.0:54333 (CTRL + C to quit)
2021-07-14T17:46:15:INFO:hypercorn.error: Running on https://0.0.0.0:4566 (CTRL + C to quit)
Waiting for all LocalStack services to be ready
Waiting for all LocalStack services to be ready
Ready.
2021-07-14T17:46:26:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 10599.67ms

2021-07-14T17:47:17:WARNING:localstack.utils.server.http2_server: Error in proxy handler for request GET http://localhost:4566/<path removed>: An error occurred (404) when calling the HeadObject operation: Not Found Traceback (most recent call last):
  File "/opt/code/localstack/localstack/utils/server/http2_server.py", line 131, in index
    raise result
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 777, in run
    result = self.func(self.params, **kwargs)
  File "/opt/code/localstack/localstack/utils/async_utils.py", line 30, in _run
    return fn(*args, **kwargs)
  File "/opt/code/localstack/localstack/services/generic_proxy.py", line 525, in handler
    server_address=parsed_url.netloc,
  File "/opt/code/localstack/localstack/services/generic_proxy.py", line 310, in modify_and_forward
    method=method, path=path, data=data, headers=headers
  File "/opt/code/localstack/localstack/services/edge.py", line 157, in forward_request
    return do_forward_request(api, method, path, data, headers, port=port)
  File "/opt/code/localstack/localstack/services/edge.py", line 185, in do_forward_request
    result = do_forward_request_inmem(api, method, path, data, headers, port=port)
  File "/opt/code/localstack/localstack/services/edge.py", line 217, in do_forward_request_inmem
    server_address=server_address,
  File "/opt/code/localstack/localstack/services/generic_proxy.py", line 385, in modify_and_forward
    updated_response = update_listener.return_response(**kwargs)
  File "/opt/code/localstack/localstack/services/s3/s3_listener.py", line 1524, in return_response
    fix_range_content_type(bucket_name, path, headers, response)
  File "/opt/code/localstack/localstack/services/s3/s3_listener.py", line 636, in fix_range_content_type
    result = s3_client.head_object(Bucket=bucket_name, Key=key_name)
  File "/opt/code/localstack/.venv/lib/python3.7/site-packages/botocore/client.py", line 386, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/opt/code/localstack/.venv/lib/python3.7/site-packages/botocore/client.py", line 705, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (404) when calling the HeadObject operation: Not Found
`
